### PR TITLE
[10_6_X] ONNXRuntime-based implementation of DeepJet, DeepAK8 and DeepDoubleX

### PR DESCRIPTION
**Pull request by hqucms**
#### PR description:

Backport of https://github.com/cms-sw/cmssw/pull/28112.

This PR adds the ONNXRuntime-based implementation of DeepJet, DeepAK8 and DeepDoubleX which provides significant speed-up. The previous TensorFlow/MXNet implementations are removed. The new ONNXRuntime-based implementation is enabled by default (as discussed in https://github.com/cms-sw/cmssw/pull/30123#issuecomment-645363153).

**Needs:**
- https://github.com/cms-sw/cmsdist/pull/5902
- https://github.com/cms-data/RecoBTag-Combined/pull/25 (already in 10_6_X)

### CODE REVIEW

**Comment by perrotta** [comment]
backport of #28112

**Comment by mrodozov** [comment]
test parameters:
- pull_requests = https://github.com/cms-sw/cmsdist/pull/5902

**Comment by mrodozov** [comment]
please test

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by hqucms** [comment]
The PR is ready for review now -- https://github.com/cms-sw/cmssw/pull/30123/commits/c9e4237437e1b9b27a05dca4eb1011deb5a77a5f puts the new ONNXRuntime-based implementations under the `run2_miniAOD_devel` modifier and keeps the old TF/MXNet implementations as the default.

**Comment by perrotta** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
The "backport" is not really trivial and the solution in config files to support the old and new inference method somewhat questionable or confusing, IMHO.
A more straightforward implementation would need all replicated variants in configuration to come with a TF/MX/ONNX suffix where appropriate with a common (RecoBTag/Combined?) area to contain the disambiguation load. This mainly refers to the part added in https://github.com/cms-sw/cmssw/pull/30123/files/5fae2a7c11d4eb401a35d7e3dac30b77b0b36019..c9e4237437e1b9b27a05dca4eb1011deb5a77a5f
Even more appropriate would be to use the common C++ part as in #28112.

Considering that the differences between TF/MX and ONNX are at numerical precision, as my first preference, I'd argue for an exception to the no-change policy so that #28112 can be backported directly.
This will have to be agreed with release planning and PPD.
@silviodonato what do you think ?

If the exception to the no-change policy is not acceptable,
Is there a really good motivation for this PR?
I think that for reminiAOD it should be acceptable to stay with the TF/MXNet implementations. IIRC, the extra CPU cost of DeepFlavor/DeepJet, DeepAK8 and DeepDoubleX is under 10% of the total miniAOD processing time.

If some of the inference has to be rerun in analysis or in nanoAOD where the CPU speedup can be vital, I would propose (as a second preference) to port just the c++ part.

**Comment by slava77** [comment]
Correction to https://github.com/cms-sw/cmssw/pull/30123#issuecomment-644924212

> Even more appropriate would be to use the common C++ part as in #29172.

File: RecoBTag/ONNXRuntime/python/pfDeepBoostedJet_cff.py [hunk 1]
```diff
@@ -1,28 +1,48 @@
...
     param_path = 'RecoBTag/Combined/data/DeepBoostedJet/V02/decorrelated/resnet-0000.params',
     debugMode = False, # debug
 )
 
+# nominal DeepAK8 (ONNXRuntime)
+run2_miniAOD_devel.toReplaceWith(pfDeepBoostedJetTags,
+    _pfDeepBoostedJetTags.clone(
+        preprocessParams = pfDeepBoostedJetPreprocessParams,
+        model_path = 'RecoBTag/Combined/data/DeepBoostedJet/V02/full/resnet.onnx',
+        debugMode  = False, # debug
```
[the remark this answers is not in the document]
Comment (silviodonato) [reply RecoBTag/ONNXRuntime/python/pfDeepBoostedJet_cff.py:33]: I'm confused... It seems that when we apply `run2_miniAOD_devel` we get the old behaviour 
```
pfDeepBoostedJetTags = _pfDeepBoostedJetTags.clone(
    preprocessParams = pfDeepBoostedJetPreprocessParams,
    model_path = 'RecoBTag/Combined/data/DeepBoostedJet/V02/full/resnet.onnx',
)
```
and you are changing the default values of `pfDeepBoostedJetTags`

**Review by silviodonato — COMMENTED** [review]
Please check the implementation of `run2_miniAOD_devel` for `pfMassDecorrelatedDeepBoostedJetTags` and `pfDeepBoostedJetTags`

**Comment by silviodonato** [comment]
> The "backport" is not really trivial and the solution in config files to support the old and new inference method somewhat questionable or confusing, IMHO.
> A more straightforward implementation would need all replicated variants in configuration to come with a TF/MX/ONNX suffix where appropriate with a common (RecoBTag/Combined?) area to contain the disambiguation load. This mainly refers to the part added in https://github.com/cms-sw/cmssw/pull/30123/files/5fae2a7c11d4eb401a35d7e3dac30b77b0b36019..c9e4237437e1b9b27a05dca4eb1011deb5a77a5f
> Even more appropriate would be to use the common C++ part as in ~#28112~ #29172.
> 
> Considering that the differences between TF/MX and ONNX are at numerical precision, as my first preference, I'd argue for an exception to the no-change policy so that #28112 can be backported directly.
> This will have to be agreed with release planning and PPD.
> @silviodonato what do you think ?
> 
> If the exception to the no-change policy is not acceptable,
> Is there a really good motivation for this PR?
> I think that for reminiAOD it should be acceptable to stay with the TF/MXNet implementations. IIRC, the extra CPU cost of DeepFlavor/DeepJet, DeepAK8 and DeepDoubleX is under 10% of the total miniAOD processing time.
> 
> If some of the inference has to be rerun in analysis or in nanoAOD where the CPU speedup can be vital, I would propose (as a second preference) to port just the c++ part.

@slava77 if "the differences between TF/MX and ONNX are at numerical precision", I agree to backport directly #28112.

**Comment by slava77** [comment]
> @slava77 if "the differences between TF/MX and ONNX are at numerical precision", I agree to backport directly #28112.

Thank you!
This PR should then become much easier.
@hqucms please update to match #28112 more closely (adding more technical pieces that are already in from   #28779, #29989,  #29692, #29581 should be OK)

**Comment by slava77** [comment]
> > @slava77 if "the differences between TF/MX and ONNX are at numerical precision", I agree to backport directly #28112.
> 
> Thank you!
> This PR should then become much easier.
> @hqucms please update to match #28112 more closely (adding more technical pieces that are already in from #28779, #29989, #29692, #29581 should be OK)

ping
@hqucms 
please clarify on the status of updating this PR.
Thank you.

**Comment by hqucms** [comment]
@slava77 I updated the PR to better match https://github.com/cms-sw/cmssw/pull/28112. Please let me know what you think!

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by mrodozov** [comment]
abort

**Comment by mrodozov** [comment]
please test

**Comment by santocch** [comment]
+1

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/30123	9671b78 
- this backports  #28112 (the large part) together with more technical updates from #28779, #29989, #29692, #29581; the TF and MXNet based inference is not moved to ONNXRuntime as in the master version. The implementation of inference is logically the same with some numerical precision differences.
    - based on the discussion leading to https://github.com/cms-sw/cmssw/pull/30123#issuecomment-645363153, the differences at the numerical precision level are acceptable
- jenkins tests pass and comparisons with the baseline show a few single-bin small differences in the b-tag discriminant values.

**Comment by peruzzim** [comment]
+xpog

**Comment by peruzzim** [comment]
@hqucms please correct the initial description to make it clear that this is not conditional on the `run2_miniAOD_devel` modifier anymore

**Comment by hqucms** [comment]
> @hqucms please correct the initial description to make it clear that this is not conditional on the `run2_miniAOD_devel` modifier anymore

@peruzzim Done!

**Comment by silviodonato** [comment]
+1

**Pull request merged**